src/topp/DatabaseSuitability.cpp

```diff
@@ -0,0 +1,367 @@
+    IdXMLFile x;
+    vector<ProteinIdentification> prot_ids;
+    vector<PeptideIdentification> pep_ids;
+    x.load(in_id, prot_ids, pep_ids);
+
+    vector<ProteinIdentification> novo_prots;
+    vector<PeptideIdentification> novo_peps;
+    x.load(in_novo, novo_prots, novo_peps);
+
+    UInt64 count_ms2_lvl;
+    {
+      MzMLFile m;
+      PeakFileOptions op;
+      op.addMSLevel(2);
+      op.setFillData(false);
+      m.setOptions(op);
+      PeakMap exp;
+      m.load(in_spec, exp);
+      count_ms2_lvl = exp.size();
+    }
+
+    //-------------------------------------------------------------
+    // calculations
+    //-------------------------------------------------------------
+
+    // db suitability
+
+
+    double cut_off;
+    if (!no_re_rank)
+    {
+      cut_off = getDecoyCutOff_(pep_ids, novo_fract);
+      if (cut_off < 0)
+      {
+        OPENMS_LOG_ERROR << "Could not compute decoy cut off. Re-ranking impossible. If you want to ignore this, set the 'force_no_re_rank' flag." << endl;
+        return INCOMPATIBLE_INPUT_DATA;
+      }
+    }
+
+    UInt64 count_db = 0;
+    UInt64 count_novo = 0;
+    UInt64 count_re_ranked = 0;
+    UInt64 count_interest = 0;
+
+    for (const auto& pep_id : pep_ids)
+    {
+      vector<PeptideHit> hits = pep_id.getHits();
+
+      if (hits.empty()) continue;
+
+      PeptideHit top_hit = hits[0];
+
+      if (top_hit.getMetaValue("target_decoy") == "decoy") continue;
+
+      // check if top hit is found in de novo protein
+      set<String> accessions = top_hit.extractProteinAccessionsSet();
+      bool is_novo = true;
+      for (const String& acc : accessions)
+      {
+        if (acc.find(Constants::UserParam::CONCAT_PEPTIDE) == String::npos)
+        {
+          is_novo = false;
+          break;
+        }
+      }
+
+      if (is_novo) // top hit is de novo hit
+      {
+        if (hits.size() == 1)
+        {
+          ++count_novo;
+          continue;
+        }
+
+        PeptideHit second_hit = hits[1];
+
+        if (second_hit.getMetaValue("target_decoy") == "decoy")
+        {
+          ++count_novo;
+          continue;
+        }
+
+        // check if second hit is db hit
+        set<String> second_accessions = top_hit.extractProteinAccessionsSet();
+        bool is_novo_too = true;
+        for (const String& acc2 : second_accessions)
+        {
+          if (acc2.find(Constants::UserParam::CONCAT_PEPTIDE) == String::npos)
+          {
+            is_novo_too = false;
+            break;
+          }
+        }
+
+        if (is_novo_too) // second hit is also de novo hit
+        {
+          ++count_novo;
+        }
+        else // second hit is db hit
+        {
+          ++count_interest;
+          // check for re-ranking
+          if (no_re_rank)
+          {
+            ++count_novo;
+            continue;
+          }
+
+          if (double(top_hit.getMetaValue("MS:1002252")) - double(second_hit.getMetaValue("MS:1002252")) <= cut_off)
```

Comment:
the metavalue might not exist... and this will crash hard

Reply:
While calculating the decoy cut-off all hits are checked to have this meta-value. I thought it wouldn't be necessary to check for existence, but I guess it would be safer to still check..